### ISSUE TITLE
jitclass array attribute not updated by looping with prange

### ISSUE THREAD

**Issue by NicolasHug** [numba/numba#3459]
## Reporting a bug

- [X] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

(Please feel free to update the title with a more accurate description!)

Running the following code will not update the array `param.a`

```python
import numpy as np
from numba import njit, jitclass, uint32, prange

size = 8

@jitclass([
    ('a', uint32[::1]),
])
class Param:
    def __init__(self):
        self.a = np.empty(size, dtype=np.uint32)

@njit(parallel=True)
def g(param):
    n_threads = 4
    # thread 0 will handle indices 0-1, thread 1 will handle indices 2-3, etc.
    for thread_idx in prange(n_threads):
        start, stop = thread_idx * 2, (thread_idx + 1) * 2
        for i in range(start, stop):
            param.a[i] = i
    print(param.a)  # should be [0, 1, 2, 3, 4, 5, 6, 7] but contains garbage

g(Param())
```

I get the expected result with `parallel=False`.

Declaring local variables will also work (or not passing a parameter at all):

```python
@njit(parallel=True)
def g(param):
    n_threads = 4
    a = param.a  # <- local variable
    # thread 0 will handle indices 0-1, thread 1 will handle indices 2-3, etc.
    for thread_idx in prange(n_threads):
        start, stop = thread_idx * 2, (thread_idx + 1) * 2
        for i in range(start, stop):
            a[i] = i
    print(param.a)  # Now OK: [0, 1, 2, 3, 4, 5, 6, 7]

```

**Comment by stuartarchibald** [comment]
Thanks for the report. The interaction between ParallelAccelerator (`parallel=True`) and `jitclass` hasn't been hugely explored. I would imagine your best bet to get things working right now is to proceed with the workarounds noted.

**Comment by stuartarchibald** [comment]
Ping @DrTodd13 @ehsantn

**Comment by JeppeDruedahl** [comment]
The workaround with declaring the variable locally does not seem to always work. 

```python
@njit(parallel=True)
def g(param):
    a = param.a
    for i in prange(size):
        a[i] = i
    print(param.a)  # should be [0, 1, 2, 3, 4, 5, 6, 7] but contains garbage
```

**Comment by DrTodd13** [comment]
@ehsantn @sklam The problem is in dead code elimination.  The "param.a[i] = i" in the loop is eliminated because param.a is not read after that point and param.a is not recognized as an argument to the function.  I am thinking a solution is that we say that param.a aliases param and then the dead code elimination code would recognize param.a as an alias for an argument and then not eliminate it.  In general, if X is an object, then X.Y is a part of X and so there is some overlap that we could call aliasing.  Would doing so eliminate too many useful optimizations?  Right now, in ir_utils.py:705, there is a similar check but it only looks at the attrs T, ctypes, and flat.

**Comment by stuartarchibald** [comment]
Seemed to be fixed in 0.45.
